src/Functions/h3PolygonToCells.cpp

```diff
@@ -0,0 +1,187 @@
+        dst_offsets.resize(input_rows_count);
+        auto current_offset = 0;
+
+
+        callOnGeometryDataType<CartesianPoint>(arguments[0].type, [&] (const auto & type)
+        {
+            using TypeConverter = std::decay_t<decltype(type)>;
+            using Converter = typename TypeConverter::Type;
+
+            // polygonToCells does not work for points and lines
+            if constexpr (std::is_same_v<ColumnToPointsConverter<CartesianPoint>, Converter>)
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The second argument of function {} must not be Point", getName());
+            else if constexpr (std::is_same_v<ColumnToLineStringsConverter<CartesianPoint>, Converter>)
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The second argument of function {} must not be LineString", getName());
+
+            // all geometries will be of same kind
+            auto geometries = Converter::convert(arguments[0].column);
+            for (size_t row = 0; row < input_rows_count; ++row)
+            {
+                // check resolution *once* before looping over polygons
+                const UInt8 resolution = data_resolution[row];
+                if (resolution > MAX_H3_RES)
+                    throw Exception(
+                        ErrorCodes::ARGUMENT_OUT_OF_BOUND,
+                        "The argument 'resolution' ({}) of function {} is out of bounds because the maximum resolution in H3 library is {}",
+                        toString(resolution), getName(), toString(MAX_H3_RES));
+
+                // will be ring, polygon, or multipolygon (boost::geometry)
+                auto geometry = geometries[row];
+                boost::geometry::correct(geometry); // TODO: Is this good or bad?
+
+                // always work on multipolygon
+                CartesianMultiPolygon multi_polygon;
+                if constexpr (std::is_same_v<ColumnToMultiPolygonsConverter<CartesianPoint>, Converter>)
+                    multi_polygon = geometry;
+                else if constexpr (std::is_same_v<ColumnToPolygonsConverter<CartesianPoint>, Converter>)
+                    multi_polygon = boost::geometry::model::multi_polygon({ geometry });
+                else if constexpr (std::is_same_v<ColumnToRingsConverter<CartesianPoint>, Converter>)
+                    multi_polygon = boost::geometry::model::multi_polygon({ boost::geometry::model::polygon({ geometry }) });
+
+                // will only iterate once for ring and polygon
+                for (auto & polygon : multi_polygon)
+                {
+                    GeoPolygon h3_polygon{};
+                    parseGeoLoopFromRing(polygon.outer(), &h3_polygon.geoloop);
+
+                    auto holes = polygon.inners();
+                    int count = static_cast<int>(std::size(holes));
+                    h3_polygon.holes = static_cast<GeoLoop*>(malloc(count * sizeof(GeoLoop)));
```

Comment:
We should avoid using `malloc`/`new` in the codebase.

Reply:
What would be the proper way to do this? I need to build a H3 `GeoPolygon` and provide it to `polygonToCells`.

Reply:
Ok I think I've found a way to avoid `malloc`ing the C structs, using wrapper classes and `std::vector` instead.